crates/meta/src/storage.rs

```diff
@@ -53,10 +76,74 @@ pub struct FileMetaStorage {
 }
 
 impl FileMetaStorage {
-    pub fn new(root_path: PathBuf) -> Self {
-        Self {
+    pub fn new(root_path: PathBuf) -> Result<Self, BuildError> {
+        if Self::is_empty_directory(root_path.as_path()) {
+            Self::write_storage_format_version_to_file(
+                root_path.as_path(),
+                STORAGE_FORMAT_VERSION,
+            )?;
+        } else {
+            Self::assert_compatible_storage_format_version(root_path.as_path())?;
+        }
+
+        Ok(Self {
             root_path,
             next_file_index: 0,
+        })
+    }
+
+    fn is_empty_directory(path: impl AsRef<Path>) -> bool {
+        let path = path.as_ref();
+
+        !path.exists()
+            || path
+                .read_dir()
+                .expect("meta storage directory must exist")
+                .count()
+                == 0
+    }
+
+    fn write_storage_format_version_to_file(
+        root_path: impl AsRef<Path>,
+        version: StorageFormatVersionType,
+    ) -> Result<(), io::Error> {
+        let root_path = root_path.as_ref();
+
+        // make sure that the root directory exists
+        std::fs::create_dir_all(root_path)?;
+
+        let version_file_path = root_path.join(STORAGE_FORMAT_VERSION_FILE_NAME);
+        assert!(
+            !version_file_path.exists(),
+            "must never overwrite an existing version file"
+        );
+
+        let version_file = std::fs::File::create(version_file_path)?;
+
+        // use a human readable format
+        serde_json::to_writer(version_file, &version)?;
+
+        Ok(())
+    }
+
+    fn assert_compatible_storage_format_version(
+        root_path: impl AsRef<Path>,
+    ) -> Result<(), BuildError> {
+        let version_file =
+            std::fs::File::open(root_path.as_ref().join(STORAGE_FORMAT_VERSION_FILE_NAME));
+
+        let version = if let Ok(version_file) = version_file {
+            serde_json::from_reader(version_file)?
+        } else {
+            // File does not exist, this indicates that the data has been written with a Restate
+            // version that does not write a version file (version w/o versioning information).
+            0
```

Comment:
I guess this in future becomes an hard error?

Reply:
Once we are certain that nobody runs with a meta storage from Restate `0.7`, then we can make it a hard error, yes.